pkg/ingester/circuitbreaker_test.go

```diff
@@ -0,0 +1,686 @@
+		# TYPE cortex_ingester_circuit_breaker_results_total counter
+		cortex_ingester_circuit_breaker_results_total{ingester="ingester",result="success"} 1
+	`
+	assert.NoError(t, testutil.GatherAndCompare(registry, strings.NewReader(expectedMetrics), metricNames...))
+}
+
+func TestCircuitBreaker_RecordError(t *testing.T) {
+	registry := prometheus.NewRegistry()
+	metricNames := []string{
+		"cortex_ingester_circuit_breaker_results_total",
+	}
+	cfg := CircuitBreakerConfig{Enabled: true, CooldownPeriod: 10 * time.Second}
+	cb := newCircuitBreaker(cfg, "ingester", log.NewNopLogger(), registry)
+	cb.recordError(context.DeadlineExceeded)
+	expectedMetrics := `
+		# HELP cortex_ingester_circuit_breaker_results_total Results of executing requests via the circuit breaker.
+		# TYPE cortex_ingester_circuit_breaker_results_total counter
+		cortex_ingester_circuit_breaker_results_total{ingester="ingester",result="error"} 1
+	`
+	assert.NoError(t, testutil.GatherAndCompare(registry, strings.NewReader(expectedMetrics), metricNames...))
+}
+
+func TestCircuitBreaker_FinishPushRequest(t *testing.T) {
+	metricNames := []string{
+		"cortex_ingester_circuit_breaker_results_total",
+	}
+	cfg := CircuitBreakerConfig{
+		Enabled:     true,
+		PushTimeout: 2 * time.Second,
+	}
+	testCases := map[string]struct {
+		delay           time.Duration
+		err             error
+		expectedErr     error
+		expectedMetrics string
+	}{
+		"delay lower than PushTimeout and no input err give success": {
+			delay: 1 * time.Second,
+			err:   nil,
+			expectedMetrics: `
+				# HELP cortex_ingester_circuit_breaker_results_total Results of executing requests via the circuit breaker.
+				# TYPE cortex_ingester_circuit_breaker_results_total counter
+				cortex_ingester_circuit_breaker_results_total{ingester="ingester",result="success"} 1
+			`,
+		},
+		"delay higher than PushTimeout and no input error give context deadline exceeded error": {
+			delay:       3 * time.Second,
+			err:         nil,
+			expectedErr: context.DeadlineExceeded,
+			expectedMetrics: `
+				# HELP cortex_ingester_circuit_breaker_results_total Results of executing requests via the circuit breaker.
+				# TYPE cortex_ingester_circuit_breaker_results_total counter
+				cortex_ingester_circuit_breaker_results_total{ingester="ingester",result="error"} 1
+			`,
+		},
+		"delay higher than PushTimeout and an input error different from context deadline exceeded give context deadline exceeded error": {
+			delay:       3 * time.Second,
+			err:         newInstanceLimitReachedError("error"),
+			expectedErr: context.DeadlineExceeded,
+			expectedMetrics: `
+				# HELP cortex_ingester_circuit_breaker_results_total Results of executing requests via the circuit breaker.
+				# TYPE cortex_ingester_circuit_breaker_results_total counter
+				cortex_ingester_circuit_breaker_results_total{ingester="ingester",result="error"} 1
+			`,
+		},
+	}
+	for testName, testCase := range testCases {
+		registry := prometheus.NewRegistry()
```

Comment:
By reusing `registry` and `cb` between tests, we may be getting metrics from previous test (esp. `result="error"`)

Reply:
Thank you for pointing this out. This wasn't intended to be put here, but inside each test.